onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -647,8 +647,10 @@ def _get_data(self, query, fields, sort, start, limit, is_public_request):
             if should_paginate:
                 self.paginator.page_size = retrieval_threshold
 
-        if not isinstance(self.object_list, types.GeneratorType) and \
-                should_paginate:
+        if isinstance(self.object_list, types.GeneratorType):
+            # Unpack generator object to list
+            self.object_list = list(self.object_list)
+        elif should_paginate:
```

Comment:
What happens if it's a generator and the user asked for pagination ?

Reply:
It will mostly result to a generator object if sort query param is passed for the JSON result.
If `page` and `page_size` headers are passed, the result will be unpacked and paginated appropriately.

Reply:
Still seems like if it's a generator object the pagination headers won't be included.... Seems like it's unpacked but the pagination headers are not included